File: csharp/Structural/10-Facade/Facade/JobDispatchFacade.cs
using Facade.Subsystem;

namespace Facade.Facade;

public class JobDispatchFacade
{
    private readonly Scheduler _scheduler = new();
    private readonly HandymanAssigner _assigner = new();
    private readonly NotificationService _notifier = new();
    private readonly JobCardRegistry _registry = new();
    private readonly BillingService _billing = new();

    public void DispatchJob(
        string jobId,
        string title,
        string category,
        string clientEmail,
        bool isUrgent = false)
    {
        Console.WriteLine($"\n{new string('=', 50)}");
        Console.WriteLine($"🚀 Dispatching Job #{jobId}: {title}");
        Console.WriteLine(new string('=', 50));

        var slot = _scheduler.FindAvailableSlot(category);
        _scheduler.BookSlot(slot, jobId);

        var handyman = _assigner.FindQualifiedHandyman(category);
        _assigner.Assign(handyman, jobId);

        _notifier.NotifyHomeowner(jobId, slot);
        _notifier.NotifyHandyman(jobId, handyman);
        _notifier.NotifySupervisor(jobId);

        _registry.Register(jobId, title);

        var quote = _billing.GenerateQuote(category, isUrgent);
        _billing.SendQuote(jobId, clientEmail, quote);

        Console.WriteLine($"✅ Job #{jobId} dispatched successfully.\n");
    }
}

File: typescript/Structural/10-Facade/Facade/JobDispatchFacade.ts
import { BillingService } from "../Subsystem/BillingService";
import { HandymanAssigner } from "../Subsystem/HandymanAssigner";
import { JobCardRegistry } from "../Subsystem/JobCardRegistry";
import { NotificationService } from "../Subsystem/NotificationService";
import { Scheduler } from "../Subsystem/Scheduler";

export class JobDispatchFacade {
  private scheduler = new Scheduler();
  private assigner = new HandymanAssigner();
  private notifier = new NotificationService();
  private registry = new JobCardRegistry();
  private billing = new BillingService();

  dispatchJob(params: {
    jobId: string;
    title: string;
    category: string;
    clientEmail: string;
    isUrgent?: boolean;
  }): void {
    const { jobId, title, category, clientEmail, isUrgent = false } = params;

    console.log(`\n${"=".repeat(50)}`);
    console.log(`🚀 Dispatching Job #${jobId}: ${title}`);
    console.log("=".repeat(50));

    const slot = this.scheduler.findAvailableSlot(category);
    this.scheduler.bookSlot(slot, jobId);

    const handyman = this.assigner.findQualifiedHandyman(category);
    this.assigner.assign(handyman, jobId);

    this.notifier.notifyHomeowner(jobId, slot);
    this.notifier.notifyHandyman(jobId, handyman);
    this.notifier.notifySupervisor(jobId);

    this.registry.register(jobId, title);

    const quote = this.billing.generateQuote(category, isUrgent);
    this.billing.sendQuote(jobId, clientEmail, quote);

    console.log(`✅ Job #${jobId} dispatched successfully.\n`);
  }
}
